### PR TITLE
[Form] Added ResetInterface to CachingFactoryDecorator

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29259
| License       | MIT
| Doc PR        | ~

I don't know if this should be considered a feature and target master, but I tend to see it as a bug fix.

### CODE REVIEW

**Review by ogizanagi — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Shouldn't we tag the corresponding service as `kernel.reset`N

**Comment by HeahDude** [comment]
Done too. Thanks!

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml [hunk 1]
```diff
@@ -57,6 +57,7 @@
 
         <service id="form.choice_list_factory.cached" class="Symfony\Component\Form\ChoiceList\Factory\CachingFactoryDecorator">
             <argument type="service" id="form.choice_list_factory.property_access"/>
+            <tag name="kernel.reset" method="reset" />
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: To preserve cross deps, the tag should be removed in FrameworkExtension when the method doesn't exist.

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml (hunk 1 above)
Comment (HeahDude) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: Not sure we have to care about that, there is a conflict https://github.com/symfony/symfony/blob/4.2/src/Symfony/Bundle/FrameworkBundle/composer.json#L71.

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: The conflict won't prevent installing 4.2.0, where the method doesnt exist.

Comment (HeahDude) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: Understood, done! Thanks

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -410,6 +411,11 @@ private function registerFormConfiguration(array $config, ContainerBuilder $cont
         if (!class_exists(Translator::class)) {
             $container->removeDefinition('form.type_extension.upload.validator');
         }
+        if (!method_exists(CachingFactoryDecorator::class, 'reset')) {
```
Comment (HeahDude) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: So I guess this can be reverted when merged in master.

**Review by xabbuh — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @HeahDude.

**Pull request merged**